mojoshader_sdlgpu.c

```diff
@@ -349,23 +378,86 @@ MOJOSHADER_sdlShaderData *MOJOSHADER_sdlCompileShader(
+
     if ((vshader == NULL) || (pshader == NULL)) /* Both shaders MUST exist! */
         return NULL;
 
-    result = (MOJOSHADER_sdlProgram*) ctx->malloc_fn(sizeof(MOJOSHADER_sdlProgram), ctx->malloc_data);
+    if (ctx->linker_cache == NULL)
+    {
+        ctx->linker_cache = hash_create(NULL, hash_shaders, match_shaders,
+                                        nuke_shaders, 0, ctx->malloc_fn,
+                                        ctx->free_fn, ctx->malloc_data);
+
+        if (ctx->linker_cache == NULL)
+        {
+            out_of_memory();
+            return NULL;
+        } // if
+    } // if
+
+    LinkedShaderData shaders;
+    shaders.vertex = vshader;
+    shaders.fragment = pshader;
+    memset(shaders.vertexAttributes, 0, sizeof(MOJOSHADER_sdlVertexAttribute) * 16);
+    shaders.vertexAttributeCount = vertexAttributeCount;
+    for (int i = 0; i < vertexAttributeCount; i += 1)
+    {
+        shaders.vertexAttributes[i] = vertexAttributes[i];
+    }
+
+    const void *val = NULL;
+
+    if (hash_find(ctx->linker_cache, &shaders, &val))
+    {
+        ctx->bound_program = (MOJOSHADER_sdlProgram *) val;
+        return ctx->bound_program;
+    }
+
+    program = (MOJOSHADER_sdlProgram*) ctx->malloc_fn(sizeof(MOJOSHADER_sdlProgram), ctx->malloc_data);
 
-    if (result == NULL)
+    if (program == NULL)
     {
         out_of_memory();
         return NULL;
     } // if
 
+	// We have to patch the SPIR-V output for non-float inputs. These are:
+	// BYTE4  - 5
+	// SHORT2 - 6
+	// SHORT4 - 7
+	int vDataLen = vshader->parseData->output_len - sizeof(SpirvPatchTable);
+	SpirvPatchTable *vTable = (SpirvPatchTable *) &vshader->parseData->output[vDataLen];
+
+	for (int i = 0; i < vertexAttributeCount; i += 1)
+	{
+		MOJOSHADER_sdlVertexAttribute *element = &vertexAttributes[i];
+		if (element->vertexElementFormat >= 5 && element->vertexElementFormat <= 7)
+		{
+			uint32 typeDecl = element->vertexElementFormat == 5 ? vTable->tid_uvec4_p : vTable->tid_ivec4_p;
+			uint32 typeLoad = element->vertexElementFormat == 5 ? vTable->tid_uvec4 : vTable->tid_ivec4;
+			SpvOp opcodeLoad = element->vertexElementFormat == 5 ? SpvOpConvertUToF : SpvOpConvertSToF;
+
+			uint32_t typeDeclOffset = vTable->attrib_type_offsets[element->usage][element->usageIndex];
+			((uint32_t*)vshader->parseData->output)[typeDeclOffset] = typeDecl;
+			for (uint32_t j = 0; j < vTable->attrib_type_load_offsets[element->usage][element->usageIndex].num_loads; j += 1)
+			{
+				uint32_t typeLoadOffset = vTable->attrib_type_load_offsets[element->usage][element->usageIndex].load_types[j];
+				uint32_t opcodeLoadOffset = vTable->attrib_type_load_offsets[element->usage][element->usageIndex].load_opcodes[j];
+				uint32_t *ptr_to_opcode_u32 = &((uint32_t*)vshader->parseData->output)[opcodeLoadOffset];
+				((uint32_t*)vshader->parseData->output)[typeLoadOffset] = typeLoad;
+				*ptr_to_opcode_u32 = (*ptr_to_opcode_u32 & 0xFFFF0000) | opcodeLoad;
+			}
+		}
```

Comment:
We should also patch in the opposite direction, in case one layout uses short and another uses float (unlikely but I wouldn't want to debug that without advance knowledge...)